src/gax/src/error/credentials.rs

```diff
@@ -0,0 +1,215 @@
+/// [Credential]: https://docs.rs/google-cloud-auth/latest/google_cloud_auth/credentials/struct.Credential.html
+#[derive(Clone, Debug)]
+pub struct CredentialError {
+    /// A boolean value indicating whether the error is retryable.
+    ///
+    /// If `true`, the operation that resulted in this error might succeed upon
+    /// retry.
+    is_retryable: bool,
+
+    /// The underlying source of the error.
+    ///
+    /// This provides more specific information about the cause of the failure.
+    source: CredentialErrorImpl,
+}
+
+#[derive(Clone, Debug)]
+enum CredentialErrorImpl {
+    SimpleMessage(String),
+    Source(Arc<dyn Error + Send + Sync>),
+}
+
+impl CredentialError {
+    /// Creates a new `CredentialError`.
+    ///
+    /// This function is only intended for use in the client libraries
+    /// implementation. Application may use this in mocks, though we do not
+    /// recommend that your write tests for specific error cases. Most tests
+    /// should use the generic type [Error][crate::error::Error] type.
+    ///
+    /// # Example
+    /// ```
+    /// # use google_cloud_gax::error::CredentialError;
+    /// # use google_cloud_gax::error::Error;
+    /// let err = CredentialError::new(
+    ///     false, Error::other("simulated non-retryable error while trying to create credentials"));
+    /// assert!(!err.is_retryable());
+    /// assert!(format!("{err}").contains("simulated non-retryable error"));
+    /// ```
+    /// # Arguments
+    /// * `is_retryable` - A boolean indicating whether the error is retryable.
+    /// * `source` - The underlying error that caused the auth failure.
+    pub fn new<T: Error + Send + Sync + 'static>(is_retryable: bool, source: T) -> Self {
+        CredentialError {
+            is_retryable,
+            source: CredentialErrorImpl::Source(Arc::new(source)),
+        }
+    }
+
+    /// Creates a new `CredentialError`.
+    ///
+    /// This function is only intended for use in the client libraries
+    /// implementation. Application may use this in mocks, though we do not
+    /// recommend that your write tests for specific error cases. Most tests
+    /// should use the generic type [Error][crate::error::Error] type.
+    ///
+    /// # Example
+    /// ```
+    /// # use google_cloud_gax::error::CredentialError;
+    /// let err = CredentialError::from_str(
+    ///     true, "simulated retryable error while trying to create credentials");
+    /// assert!(err.is_retryable());
+    /// assert!(format!("{err}").contains("simulated retryable error"));
+    /// ```
+    ///
+    /// # Arguments
+    /// * `is_retryable` - A boolean indicating whether the error is retryable.
+    /// * `message` - The underlying error that caused the auth failure.
+    pub fn from_str<T: Into<String>>(is_retryable: bool, message: T) -> Self {
+        CredentialError::new(
+            is_retryable,
+            CredentialErrorImpl::SimpleMessage(message.into()),
+        )
```

Comment:
This seems a bit strange to me. It means that `self.source` never contains the `SimpleMessage` branch. 

I am going to open a separate bug to clean this up, for now I just want to copy the code from `auth`.

Reply:
> This seems a bit strange to me. It means that `self.source` never contains the `SimpleMessage` branch.

Huh? what makes you say that?

https://github.com/googleapis/google-cloud-rust/blob/099c37c5799747a65e58b2a20115652f26a20e3f/src/gax/src/error/credentials.rs#L188-L193

Reply:
Indeed.